merge_values.go

```diff
@@ -7,6 +7,93 @@ import (
 	yaml "gopkg.in/yaml.v2"
 )
 
+// MergeValues merges config values so they can be used when installing or
+// updating Helm releases. It takes in 2 maps with a string key and YAML values
+// passed as a byte array.
+//
+// A deep merge is performed into a single map[string]interface{} output. If a
+// value is present in both then the source map is preferred.
+//
+// The YAML values are parsed using yamlToStringMap. This is because the
+// default behaviour of the YAML parser is to unmarshal into
+// map[interface{}]interface{} which causes problems with the merge logic.
+// See https://github.com/go-yaml/yaml/issues/139.
+//
+func MergeValues(destMap, srcMap map[string][]byte) (map[string]interface{}, error) {
+	result := map[string]interface{}{}
+
+	destVals, err := processYAML(destMap)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	srcVals, err := processYAML(srcMap)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	result = mergeValues(destVals, srcVals)
+
+	return result, nil
+}
+
+// processYAML accepts a map with a string key and YAML values passed as a
+// byte array. Only a single key is supported and the input data structure
+// matches the configmap or secret where the data is stored.
+func processYAML(inputMap map[string][]byte) (map[string]interface{}, error) {
+	var err error
+
+	result := map[string]interface{}{}
+
+	if len(inputMap) > 1 {
+		return nil, microerror.Maskf(executionFailedError, "merging %d keys is unsupported expected 1 key", len(inputMap))
+	}
+
+	for _, v := range inputMap {
+		result, err = yamlToStringMap(v)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	return result, nil
+}
+
+// mergeValues implements the merge logic. It performs a deep merge. If a value
+// is present in both then the source map is preferred.
+//
+// Logic is based on the upstream logic implemented by Helm.
+// https://github.com/helm/helm/blob/240e539cec44e2b746b3541529d41f4ba01e77df/cmd/helm/install.go#L358
```

Comment:
See this link for the upstream code which the merge logic is based on. I just made minor changes to follow our usual conventions.